bff/pkg/server/server.go

```diff
@@ -513,6 +521,34 @@ func activity_handler(metrics_service metrics.Service, validate *validator.Valid
 			return
 		}
 
+		// populate query_fp with query text
+		for _, result := range results {
+			switch metric := result["metric"].(type) {
+			case map[string]interface{}:
+				if queryFP, ok := metric["query_fp"].(string); ok {
+					queryText, err := query_storage.GetQuery(queryFP)
+					if err != nil {
+						http.Error(w, "Error fetching query text: "+err.Error(), http.StatusInternalServerError)
+						return
+					}
+					metric["query_fp"] = queryText
+					result["metric"] = metric
+				}
+			case map[string]string:
+				if queryFP, ok := metric["query_fp"]; ok {
+					queryText, err := query_storage.GetQuery(queryFP)
+					if err != nil {
+						http.Error(w, "Error fetching query text: "+err.Error(), http.StatusInternalServerError)
+						return
+					}
+					metric["query_fp"] = queryText
+					result["metric"] = metric
+				}
```

Comment:
i couldnt see a useful way to avoid duplication here. Because we have to type check metric before we can extract the query_fp value from it. And theres no way to check both possible types in a single statement. So if we made a function to perform that operation, that function would have the same code duplication inside it anyways, changing nothing.

Reply:
The code inside `if` is the exactly the same with the same types. You can extract that as a function.

Reply:
Basically any code that deals with the metric object has to  be duplicated because you have to work with it as a specific type in order to extract data from it or put data into it. 
Inside the if statement there are 4 statements, 2 of them involve metric, and that just leaves fetching the query and the error check. So if i extracted just those 2 statements, its basically the same thing, a new function i call and then an error check on the error it returns.

Maybe im missing something obvious but when i try to reduce this code any further it doesnt work it just moves the duplication somewhere else because in order to do anything to the metric object i have to get its specific type, of which there are 2 possible actions